src/rule_definition_tools.jl

```diff
@@ -258,3 +249,144 @@ This is able to deal with fairly complex expressions for `f`:
 propagator_name(f::Expr, propname::Symbol) = propagator_name(f.args[end], propname)
 propagator_name(fname::Symbol, propname::Symbol) = Symbol(fname, :_, propname)
 propagator_name(fname::QuoteNode, propname::Symbol) = propagator_name(fname.value, propname)
+
+"""
+    @non_differentiable(signature_expression)
+
+A helper to make it easier to declare that a method is not not differentiable.
+This is a short-hand for defining a [`frule`](@ref) and an [`rrule`](@ref)+ pullback that 
+returns [`DoesNotExist()`](@ref) for all partials (except for the function `s̄elf`-partial
+itself which is `NO_FIELDS`)
+
+The usage is to put the macro before a function signature.
+Keyword arguments should not be included.
+
+```jldoctest
+julia> @non_differentiable Base.:(==)(a, b)
+
+julia> _, pullback = rrule(==, 2.0, 3.0);
+
+julia> pullback(1.0)
+(Zero(), DoesNotExist(), DoesNotExist())
+```
+
+You can place type-constraints in the signature:
+```jldoctest
+julia> @non_differentiable Base.length(xs::Union{Number, Array})
+
+julia> frule((Zero(), 1), length, [2.0, 3.0])
+(2, DoesNotExist())
+```
```

Comment:
```suggestion
julia> frule((Zero(), 1), length, (2.0, 3.0)) # returns nothing

```
to show that the constraint has indeed been imposed correctly.
```

Reply:
Additionally, it doesn't look like there are currently tests for multi-arg functions with constraints, and mixtures of constrained and unconstrained args. Would be good to add these.

Reply:
why?
Is there a reason to expect that it would do mixtures differently?

Reply:
Haha I guess I was coming at this from the other direction -- is there a reason to assume that it _wont'_ behave differently for mixtures?

Reply:
because we assume the code is written as the simplest possible code that passes all tests.
It is simpler to treat all arguments the same than to special case 1 argument